OpenRA.Mods.Common/Traits/BotModules/SendUnitToAttackBotModule.cs

```diff
@@ -0,0 +1,183 @@
+
+		[Desc("Target types that can't be targeted.")]
+		public readonly BitSet<TargetableType> InvalidTargets;
+
+		[Desc("Should attack the furthest or closest target. Possible values are Closest, Furthest, Random")]
+		public readonly AttackDistance AttackDistance = AttackDistance.Closest;
+
+		[Desc("Attack order name.")]
+		public readonly string AttackOrderName = "Attack";
+
+		[Desc("Find target and try attack target in this interval.")]
+		public readonly int ScanTick = 463;
+
+		[Desc("The total attack desire increase this amount per scan",
+		"Note: when there is no attack unit, total attack desire will return 0.")]
+		public readonly int AttackDesireIncreasedPerScan = 10;
+
+		public override object Create(ActorInitializer init) { return new SendUnitToAttackBotModule(init.Self, this); }
+	}
+
+	public class SendUnitToAttackBotModule : ConditionalTrait<SendUnitToAttackBotModuleInfo>, IBotTick
+	{
+		readonly World world;
+		readonly Player player;
+		readonly Predicate<Actor> unitCannotBeOrdered;
+		readonly Predicate<Actor> unitCannotBeOrderedOrIsBusy;
+		readonly Predicate<Actor> isInvalidActor;
+		int minAssignRoleDelayTicks;
+		Player targetPlayer;
+		int desireIncreased;
+
+		public SendUnitToAttackBotModule(Actor self, SendUnitToAttackBotModuleInfo info)
+		: base(info)
+		{
+			world = self.World;
+			player = self.Owner;
+			isInvalidActor = a => a == null || a.IsDead || !a.IsInWorld || a.Owner != targetPlayer;
+			unitCannotBeOrdered = a => a == null || a.IsDead || !a.IsInWorld || a.Owner != player;
+			unitCannotBeOrderedOrIsBusy = a => unitCannotBeOrdered(a) || (!a.IsIdle && a.CurrentActivity is not FlyIdle);
+			desireIncreased = 0;
+		}
+
+		protected override void TraitEnabled(Actor self)
+		{
+			// Avoid all AIs reevaluating assignments on the same tick, randomize their initial evaluation delay.
+			minAssignRoleDelayTicks = world.LocalRandom.Next(0, Info.ScanTick);
+		}
+
+		void IBotTick.BotTick(IBot bot)
+		{
+			if (--minAssignRoleDelayTicks <= 0)
+			{
+				minAssignRoleDelayTicks = Info.ScanTick;
+
+				var attackdesire = 0;
+				var actors = world.ActorsWithTrait<IPositionable>().Select(at => at.Actor).Where(a =>
+				{
+					if (Info.ActorTypesAndAttackDesire.ContainsKey(a.Info.Name) && !unitCannotBeOrderedOrIsBusy(a))
+					{
+						attackdesire += Info.ActorTypesAndAttackDesire[a.Info.Name];
+						return true;
+					}
+
+					return false;
+				}).ToList();
+
+				if (actors.Count == 0)
+					desireIncreased = 0;
+				else
+					desireIncreased += Info.AttackDesireIncreasedPerScan;
+
+				if (desireIncreased + attackdesire < 100)
+					return;
+
+				// Randomly choose enemy player to attack
+				var enemyPlayers = world.Players.Where(p => p.RelationshipWith(player) == PlayerRelationship.Enemy && p.WinState != WinState.Lost).ToArray();
+				if (enemyPlayers.Length == 0)
+					return;
+
+				targetPlayer = enemyPlayers.Where(p => p.RelationshipWith(player) == PlayerRelationship.Enemy).Random(world.LocalRandom);
```

Comment:
The `Where` here seems redundant with how `enemyPlayers` is calculated.
```suggestion
				targetPlayer = enemyPlayers.Random(world.LocalRandom);
```

Reply:
fixed